grid_search.py

```diff
@@ -50,45 +58,44 @@ def grid_search_next_run(
             )
 
     # we can only deal with discrete params in a grid search
-    discrete_params = [p for p in params if p.type == HyperParameter.CATEGORICAL]
+    discrete_params = HyperParameterSet(
+        [p for p in params if p.type == HyperParameter.CATEGORICAL]
+    )
 
     # build an iterator over all combinations of param values
     param_names = [p.name for p in discrete_params]
     param_values = [p.config["values"] for p in discrete_params]
-    param_value_set = list(itertools.product(*param_values))
+    param_hashes = [
```

Comment:
Does this mean that a grid search with:
```
parameters:
  p:
    values:
      - outer_layers: 1
        inner_layers: 1
        inner_blocks: 1
      - outer_layers: 2
        inner_layers: 2
        inner_blocks: 2
      - outer_layers: 2
        inner_layers: 2
        inner_blocks: 2
```

only has 2 configurations?

Reply:
Yes, and that case is specifically tested here: 

https://github.com/wandb/sweeps/blob/master/tests/test_grid_search.py#L96

Reply:
Is this a change in behavior from "old sweeps"?
I recall people doing grid searches like:
```
p:
  values: [1,1,1,1,2]

and getting 5 configurations.. i could be mistaken, but that was my recollection of the code.

Reply:
Nope, old logic only gave 2 configurations. 

I just ran this config:

```
program: train.py
method: grid
parameters:
  p:
    values: [1,1,1,1,2]
```

against a version of core with the original anaconda1 sweep provider, here is the output:

```
(wandb-3.6) danielgoldstein@Daniels-MBP duplicates % WANDB_BASE_URL=http://api.wandb.test wandb agent dannygoldstein/debugging-duplicates/5kl8ylxu
wandb: Starting wandb agent 🕵️
2021-08-31 14:32:51,685 - wandb.wandb_agent - INFO - Running runs: []
2021-08-31 14:32:51,814 - wandb.wandb_agent - INFO - Agent received command: run
2021-08-31 14:32:51,814 - wandb.wandb_agent - INFO - Agent starting run with config:
	p: 1
2021-08-31 14:32:51,815 - wandb.wandb_agent - INFO - About to run command: /usr/bin/env python train.py --p=1
wandb: Currently logged in as: dannygoldstein (use `wandb login --relogin` to force relogin)
2021-08-31 14:32:56,823 - wandb.wandb_agent - INFO - Running runs: ['4inqjsc1']
wandb: Tracking run with wandb version 0.12.2.dev1
wandb: Syncing run feasible-sweep-1
wandb: ⭐️ View project at http://app.wandb.test/dannygoldstein/debugging-duplicates
wandb: 🧹 View sweep at http://app.wandb.test/dannygoldstein/debugging-duplicates/sweeps/5kl8ylxu
wandb: 🚀 View run at http://app.wandb.test/dannygoldstein/debugging-duplicates/runs/4inqjsc1
wandb: Run data is saved locally in /Users/danielgoldstein/debugging/duplicates/wandb/run-20210831_143252-4inqjsc1
wandb: Run `wandb offline` to turn off syncing.

{'p': 1}

wandb: Waiting for W&B process to finish, PID 8955
wandb: Program ended successfully.
wandb:
wandb: Find user logs for this run at: /Users/danielgoldstein/debugging/duplicates/wandb/run-20210831_143252-4inqjsc1/logs/debug.log
wandb: Find internal logs for this run at: /Users/danielgoldstein/debugging/duplicates/wandb/run-20210831_143252-4inqjsc1/logs/debug-internal.log
wandb: Synced 5 W&B file(s), 0 media file(s), 0 artifact file(s) and 0 other file(s)
wandb:
wandb: Synced feasible-sweep-1: http://app.wandb.test/dannygoldstein/debugging-duplicates/runs/4inqjsc1
2021-08-31 14:33:06,902 - wandb.wandb_agent - INFO - Cleaning up finished run: 4inqjsc1
2021-08-31 14:33:07,008 - wandb.wandb_agent - INFO - Agent received command: run
2021-08-31 14:33:07,008 - wandb.wandb_agent - INFO - Agent starting run with config:
	p: 2
2021-08-31 14:33:07,009 - wandb.wandb_agent - INFO - About to run command: /usr/bin/env python train.py --p=2
wandb: Currently logged in as: dannygoldstein (use `wandb login --relogin` to force relogin)
wandb: Tracking run with wandb version 0.12.2.dev1
wandb: Syncing run prime-sweep-2
wandb: ⭐️ View project at http://app.wandb.test/dannygoldstein/debugging-duplicates
wandb: 🧹 View sweep at http://app.wandb.test/dannygoldstein/debugging-duplicates/sweeps/5kl8ylxu
wandb: 🚀 View run at http://app.wandb.test/dannygoldstein/debugging-duplicates/runs/l04omar7
wandb: Run data is saved locally in /Users/danielgoldstein/debugging/duplicates/wandb/run-20210831_143308-l04omar7
wandb: Run `wandb offline` to turn off syncing.

{'p': 2}
2021-08-31 14:33:12,018 - wandb.wandb_agent - INFO - Running runs: ['l04omar7']

wandb: Waiting for W&B process to finish, PID 9038
wandb: Program ended successfully.
wandb:
wandb: Find user logs for this run at: /Users/danielgoldstein/debugging/duplicates/wandb/run-20210831_143308-l04omar7/logs/debug.log
wandb: Find internal logs for this run at: /Users/danielgoldstein/debugging/duplicates/wandb/run-20210831_143308-l04omar7/logs/debug-internal.log
wandb: Synced 5 W&B file(s), 0 media file(s), 0 artifact file(s) and 0 other file(s)
wandb:
wandb: Synced prime-sweep-2: http://app.wandb.test/dannygoldstein/debugging-duplicates/runs/l04omar7
2021-08-31 14:33:22,105 - wandb.wandb_agent - INFO - Cleaning up finished run: l04omar7
2021-08-31 14:33:22,213 - wandb.wandb_agent - INFO - Agent received command: exit
2021-08-31 14:33:22,214 - wandb.wandb_agent - INFO - Received exit command. Killing runs and quitting.
wandb: Terminating and syncing runs. Press ctrl-c to kill.
```

also, the current grid search logic on master is straight out of anaconda1, and this test is passing against that logic.

Reply:
got it.  i looked at the old code closer, i see where this ends up being de-dupped.